runtime: enable stack traces on Windows

compnerd opened the pull request:
We would not previously symbolicate the stack trace and as a result
would not display the stack trace.  Add symbolication support to the
runtime to actually make use of the captured stack trace.  This allows
us to get a stack trace when the standard library or swift code reports
a fatal error.

<!-- What's in this pull request? -->
Replace this paragraph with a description of your changes and rationale. Provide links to external references/discussions if appropriate.

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->
Resolves [SR-NNNN](https://bugs.swift.org/browse/SR-NNNN).

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

compnerd commented:
@swift-ci please test

compnerd commented:
@swift-ci please test

compnerd commented:
@mikeash - you may have some thoughts on how to handle the memory around here.  The problem is that the symbolication on Windows requires user allocation, and that means that we need to figure out how to deal with the buffers for the symbols.  For now, I've stuck in some ugly frees to deal with it, but, I am open to ideas.

compnerd commented:
@swift-ci please test

mikeash commented:
Could/should we have it accept a buffer from the caller to use, so that it doesn't have to make any heap allocations?

Alternately, have it return a `unique_ptr<char *>` so the caller doesn't have to worry about cleanup? On platforms where the string doesn't need to be allocated, it could just have a no-op deleter.

compnerd commented:
@mikeash - you can't do that!  That's my question! :-p `SymbolInfo` is a local wrapper over `dl_sym` to paper over the differences on Darwin and Linux.  So, I don't really feel too terrible about making it a horrible thing like contain a `std::string`.  But, the `unique_ptr` idea is pretty cool too and one that I didn't consider.  Do you have a preference between the two?  I think that accepting a buffer would be a really nice optimization, the question is, do we want that change to the API?

mikeash commented:
Changing the API seems fine. There aren't a ton of places that use this. You're already changing the API by requiring callers to free the result.

I think `unique_pointer` would be pretty nice. On other platforms, it will be like now: you get a pointer to static storage. On Windows, it will be like what you have here: you allocate space, then free it when done, but the freeing would happen automatically.

Having the caller pass in a buffer would let you avoid the heap allocation on Windows, but I'm not sure if it's worth the pain it would impose on the callers. I don't imagine this is too performance sensitive.

compnerd commented:
Sounds good to me!  Thanks for the input.  I'll adjust this accordingly.

compnerd committed:
runtime: enable stack traces on Windows

We would not previously symbolicate the stack trace and as a result
would not display the stack trace.  Add symbolication support to the
runtime to actually make use of the captured stack trace.  This allows
us to get a stack trace when the standard library or swift code reports
a fatal error.

compnerd commented:
@mikeash - this should probably do the trick wrt the `unique_ptr`

compnerd commented:
@swift-ci please test

compnerd commented:
@swift-ci please test

compnerd commented:
@swift-ci please test

mikeash commented:
Yay, this version looks pretty nice.

compnerd commented:
@swift-ci please test and merge

compnerd commented:
@swift-ci please test and merge
